File: kotlin-orders/kotlin-orders-dto/src/main/java/com/mvv/bank/orders/domain/Trade.kt
package com.mvv.bank.orders.domain

import com.mvv.bank.log.safe
import java.math.BigDecimal
import java.time.ZonedDateTime


interface Trade<P> {
    val market: MarketSymbol

    val product: P
    val buySellType: BuySellType
    val volume: BigDecimal // integer in case of stocks; float in case of cash/money
    val price: Amount

    val amount: Amount get() = volume * price
    val tradedAt: ZonedDateTime
}

data class FxCashTrade (
    val id: Long,

    override val market: MarketSymbol,

    override val buySellType: BuySellType,
    val buyCurrency: Currency,
    val sellCurrency: Currency,
    override val volume: BigDecimal,
    override val price: Amount,

    override val tradedAt: ZonedDateTime,
) : Trade<Currency> {

    init {
        check(price.currency == priceCurrency) {
            "Seems price ${price.safe} has wrong currency for deal ${this}." }
    }

    @Deprecated("Better to use buyCurrency/sellCurrency directly if you access/config cash trade by direct CashTrade typed variable.")
    override val product: Currency get() = when (buySellType) {
        //null -> null
        BuySellType.BUY  -> buyCurrency
        BuySellType.SELL -> sellCurrency
    }

    val priceCurrency: Currency get() = when (buySellType) {
        // opposite
        //null -> null
        BuySellType.BUY  -> sellCurrency
        BuySellType.SELL -> buyCurrency
    }
}

data class StockTrade (
    val id: Long,

    override val market: MarketSymbol,
    val company: CompanySymbol,

    override val buySellType: BuySellType,
    override val volume: BigDecimal,
    override val price: Amount,
    override val tradedAt: ZonedDateTime,
) : Trade<CompanySymbol> {
    override val product: CompanySymbol get() = company
}

File: scala-orders/src/main/scala/com/mvv/bank/orders/domain/Trade.scala
package com.mvv.bank.orders.domain

import scala.language.strictEquality
//
import java.time.ZonedDateTime
//
import com.mvv.log.safe
import com.mvv.utils.check
import com.mvv.nullables.NullableCanEqualGivens


trait Trade[P] :
  def market: MarketSymbol
  def product: P
  def buySellType: BuySellType
  def volume: BigDecimal // integer in case of stocks; float in case of cash/money
  def price: Amount
  def amount: Amount = volume * price
  def tradedAt: ZonedDateTime


case class FxCashTrade (
  id: Long,

  override val market: MarketSymbol,

  override val buySellType: BuySellType,
  buyCurrency: Currency,
  sellCurrency: Currency,
  override val volume: BigDecimal,
  override val price: Amount,

  override val tradedAt: ZonedDateTime,
  ) extends Trade[Currency] :

  check(price.currency == priceCurrency,
    s"Seems price ${price.safe} has wrong currency for deal $this.")

  //@Deprecated("Better to use buyCurrency/sellCurrency directly if you access/config cash trade by direct CashTrade typed variable.")
  override def product: Currency = buySellType match
      //null => null
    case BuySellType.BUY  => buyCurrency
    case BuySellType.SELL => sellCurrency

  def priceCurrency: Currency = buySellType match
      // opposite
      //null -> null
    case BuySellType.BUY  => sellCurrency
    case BuySellType.SELL => buyCurrency

object FxCashTrade extends NullableCanEqualGivens[FxCashTrade]


case class StockTrade (
  id: Long,

  override val market: MarketSymbol,
  company: CompanySymbol,

  override val buySellType: BuySellType,
  override val volume: BigDecimal,
  override val price: Amount,
  override val tradedAt: ZonedDateTime,
  ) extends Trade[CompanySymbol] derives CanEqual :
  override def product: CompanySymbol = company


object StockTrade extends NullableCanEqualGivens[StockTrade]
